### PR TITLE
GoogleNet

### PR DESCRIPTION
**Pull request by themurtazanazir**
GoogleNet with code

### CODE REVIEW

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 1]
```diff
@@ -4,46 +4,124 @@
...
+- A dropout layer with 70% ratio of dropped outputs.
+- A linear layer with softmax loss as the classifier.
+
+These auxiliary classifiers are removed at inference time. However, minimal gains are achieved from using auxiliary classifiers (0.5%).
+
+![](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/googlenet_auxiliary_classifier.jpg)
+
+Figure 3: An Auxiliary Classifier
+
+### 3. Architecture - GoogLeNet
```
Comment (sezan92) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:72]:
```suggestion
#### 2.4 Full Architecture - GoogLeNet
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 2]
```diff
@@ -4,46 +4,124 @@
 
 The Inception architecture, a convolutional neural network (CNN) designed for tasks in computer vision such as classification and detection, stands out due to its efficiency. It contains fewer than 7 million parameters and is significantly more compact than its predecessors, being 9 times smaller than AlexNet and 22 times smaller than VGG16. This architecture gained recognition in the ImageNet 2014 challenge, where Google's adaptation, named GoogLeNet (a tribute to LeNet), set new benchmarks in performance while utilizing fewer parameters compared to previous leading methods.
 
+
+
+### 2. Architectural Innovations
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:9]:
```suggestion
### Architectural Innovations
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 3]
```diff
@@ -4,46 +4,124 @@
...
+
+Before the advent of the Inception architecture, models like AlexNet and VGG demonstrated the benefits of deeper network structures. However, deeper networks typically entail more computational steps and can lead to issues such as overfitting and the vanishing gradient problem. The Inception architecture offers a solution, enabling the training of complex CNNs with a reduced count of floating-point parameters.
+
+#### 2.1 The Inception "Network In Network" Module
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:13]:
```suggestion
#### The Inception "Network In Network" Module 
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 4]
```diff
@@ -4,46 +4,124 @@
...
+
+#### 2.1 The Inception "Network In Network" Module 
+
+In prior networks, such as AlexNet or VGG, the fundamental block is the Convolution layer itself, but Lin et. al 2013, introduced the concept of Network In Network, arguing that a single convolution is not necessarily a correct fundamental building block. It ought to be more complex. So, inspired by that, the Inception model, decided to have a more complex building block called the Inception Module, aptly named after the famous movie - "The Inception" (dream in dream).
+
+The Inception Module insists on applying convolution filters of different kernel sizes for feature extraction at multiple scales. For any input feature map, it applies a $1 \times 1$ convolution, a 3x3 convolution, and a 5x5 convolution in parallel. In addition to convolution a max pooling operation is also applied. All four operations have padding and stride in such a way as to have the same spatial dimension. These features are concatenated and form the input to the next stage. See Figure 1.
+
+
+![](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/inception_naive.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:20]: feel free to add alt text, you could even give credit there

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 5]
```diff
@@ -4,46 +4,124 @@
...
+
+We would also want to reduce the output features of max pooling before concatenating with the output feature map. So, we add one more $1\times 1$ convolution after the maxpooling layer. We also add a ReLU activation after each $1\times 1$ convolution increasing non linearity and complexity of the module. See Figure 2.
+
+![](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/inception_reduced.png)
+
+Figure 2: Inception Module
+
+Also, because of the parallel operations of convolutions at multiple scales, we are ensuring more operations without going deeper into the network, essentially mitigating the vanishing gradient problem.
+
+#### 2.2 Average Pooling
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:36]:
```suggestion
#### Average Pooling
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 6]
```diff
@@ -4,46 +4,124 @@
...
 
-## Code 
+
+However, with an Inception block at our disposal, we do not need fully connected layers and a simple average pooling along the spatial dimensions should be enough. This was also derived from the Network in Network paper. However, GoogLeNet included one fully connected layer. They reported an increase of 0.6% in top-1 accuracy. 
+
+
+GoogLeNet has only 15% of the parameters in the fully connected layers.
+
+
+#### 2.3 Auxiliary Classifiers
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:53]:
```suggestion
#### Auxiliary Classifiers
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx (hunk 1 above)
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:72]:
```suggestion
### Architecture - GoogLeNet
```

**Review by merveenoyan — COMMENTED** [review]
thank you! mostly nits

**Comment by merveenoyan** [comment]
can you run `make quality`

**Comment by sezan92** [comment]
> can you run `make quality`

@merveenoyan done in https://github.com/johko/computer-vision-course/pull/220/commits/6afb645870f7fe2910b4a65dc2c1707f7f5bc5c3

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 7]
```diff
@@ -4,46 +4,122 @@
...
+
+We want to convert a feature map of $S \times S \times 128$ to $S \times S \times 256$ via a 5x5 convolution. The number of parameters(excluding biases) is 5x5x128x256 = 819,200. However, if we reduce the feature dimension first by a $1 \times 1$ convolution to 64, then the number of parameters(excluding biases) is $1\times 1\times 128\times 64 + 5\times 5\times 64\times 256 = 8,192 + 409,600 = 417,792$. That means the number of parameters was reduced by almost half!
+
+We would also want to reduce the output features of max pooling before concatenating with the output feature map. So, we add one more $1\times 1$ convolution after the maxpooling layer. We also add a ReLU activation after each $1\times 1$ convolution increasing non linearity and complexity of the module. See Figure 2.
+
+![](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/inception_reduced.png)
```
Comment (sezan92) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:30]:
```suggestion
![inception_reduced](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/inception_reduced.png)
```

**Review by ratan — COMMENTED** [review]
Looks Good

**Comment by ratan** [comment]
Some tests are failing, please look into it.

**Comment by themurtazanazir** [comment]
@ratan this error seems to be part of the main branch already. I can see it originating at `Unit 13 - Outlook/hyena.mdx`. I believe it is due to latex tags not being rendered in `<Tip>` tags. An issue has been raised around it as well #222.

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 8]
```diff
@@ -4,46 +4,122 @@
...
+
+These auxiliary classifiers are removed at inference time. However, minimal gains are achieved from using auxiliary classifiers (0.5%).
+
+![googlenet_aux_clf](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/googlenet_auxiliary_classifier.jpg)
+
+Figure 3: An Auxiliary Classifier
+
+### Architecture - GoogLeNet
+
+The complete architecture of GoogLeNet is shown in Figure <below>. All convolutions, including inside the inception block, use ReLU activation. It starts with two Convolution(s) + Max Pooling blocks. This is followed by a block of two inception modules (3a and 3b) followed by a max pooling. This follows a block of 5 inception blocks (4a, 4b, 4c, 4d, 4e) and a max pooling after. The auxiliary classifiers are taken out from outputs of 4a and 4d. Two inception blocks follow (5a and 5b). After this, an average pooling and a fully connected layer of 128 units are used.
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:74]:
```suggestion
The complete architecture of GoogLeNet is shown in Figure below. All convolutions, including inside the inception block, use ReLU activation. It starts with two convolution(s) and max-pooling blocks. This is followed by a block of two inception modules (3a and 3b) and a max pooling. This follows a block of 5 inception blocks (4a, 4b, 4c, 4d, 4e) and a max pooling after. The auxiliary classifiers are taken out from outputs of 4a and 4d. Two inception blocks follow (5a and 5b). After this, an average pooling and a fully connected layer of 128 units are used. 
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx [hunk 9]
```diff
@@ -4,46 +4,122 @@
...
+Figure 3: An Auxiliary Classifier
+
+### Architecture - GoogLeNet
+
+The complete architecture of GoogLeNet is shown in Figure <below>. All convolutions, including inside the inception block, use ReLU activation. It starts with two Convolution(s) + Max Pooling blocks. This is followed by a block of two inception modules (3a and 3b) followed by a max pooling. This follows a block of 5 inception blocks (4a, 4b, 4c, 4d, 4e) and a max pooling after. The auxiliary classifiers are taken out from outputs of 4a and 4d. Two inception blocks follow (5a and 5b). After this, an average pooling and a fully connected layer of 128 units are used. 
+
+![googlenet_arch](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/googlenet_architecture.png)
+Figure 4: Complete GoogLeNet Architecture
+
+### 3. Code
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx:79]:
```suggestion
### Code 
```

**Review by merveenoyan — APPROVED** [review]
Fix LaTeX and more

**Comment by themurtazanazir** [comment]
> @themurtazanazir can you incorporate following changes and the CI will go green
> 
> [8b410ff](https://github.com/johko/computer-vision-course/commit/8b410ff9b0210a0dbe86f540a040141f103e22c3) [8b410ff](https://github.com/johko/computer-vision-course/commit/8b410ff9b0210a0dbe86f540a040141f103e22c3) [bd5673b](https://github.com/johko/computer-vision-course/commit/bd5673b36b8b75a39e941a211ba10c023c349399) [63cabaa](https://github.com/johko/computer-vision-course/commit/63cabaa80138bc3128b00163980576d58a19ce8f)

@merveenoyan it seems these changes are in your PRs (#224 and #224 ). I can wait till these are merged and resync my PR with main afterwards, if there is no issue with that.

**Comment by merveenoyan** [comment]
@themurtazanazir since they're very small changes I'd rather have them fixed here and make your PR merged because my PR is quite large and will take time until it's merged

**Comment by themurtazanazir** [comment]
@merveenoyan understood. making changes.

File: chapters/en/Unit 13 - Outlook/hyena.mdx [hunk 10]
```diff
@@ -67,7 +71,7 @@ This network takes the positional index and potentially positional encodings as
 This implies that instead of learning the values of the convolution filter directly, we learn a mapping from a temporal positional encoding to the values, which is more computationally efficient, especially for long sequences.
 
 <Tip>
-It's important to note that the mapping function  $\gamma_{\theta}$  can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
+It's important to note that the mapping function  (\\ gamma_{\theta} \\)  can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```
Comment (sezan92) [inline chapters/en/Unit 13 - Outlook/hyena.mdx:74]:
```suggestion
It's important to note that the mapping function can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```

File: chapters/en/Unit 13 - Outlook/hyena.mdx [hunk 11]
```diff
@@ -67,7 +71,7 @@ This network takes the positional index and potentially positional encodings as
 This implies that instead of learning the values of the convolution filter directly, we learn a mapping from a temporal positional encoding to the values, which is more computationally efficient, especially for long sequences.
 
 <Tip>
-It's important to note that the mapping function  $\gamma_{\theta}$  can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
+It's important to note that the mapping function can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```
Comment (sezan92) [inline chapters/en/Unit 13 - Outlook/hyena.mdx:74]:
```suggestion
It's important to note that the mapping function (\\(gamma_theta\\) can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```

File: chapters/en/Unit 13 - Outlook/hyena.mdx [hunk 12]
```diff
@@ -67,7 +71,7 @@ This network takes the positional index and potentially positional encodings as
 This implies that instead of learning the values of the convolution filter directly, we learn a mapping from a temporal positional encoding to the values, which is more computationally efficient, especially for long sequences.
 
 <Tip>
-It's important to note that the mapping function  $\gamma_{\theta}$  can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
+It's important to note that the mapping function (\\(gamma_theta\\) can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```
Comment (sezan92) [inline chapters/en/Unit 13 - Outlook/hyena.mdx:74]:
```suggestion
It's important to note that the mapping function \\(gamma_theta\\) can be conceptualized within various abstract models, such Neural Field or State Space Models (S4) as discussed in [H3 paper](https://arxiv.org/abs/2212.14052).
```

**Comment by sezan92** [comment]
@MKhalusova can you check and if okay approve the PR ?

**Comment by merveenoyan** [comment]
@themurtazanazir if you can solve merge conflicts Maria will review and we can ship ⛴️

**Review by MKhalusova — APPROVED** [review]
Looks good to me, thank you for working on this!

**Comment by sezan92** [comment]
@merveenoyan @MKhalusova the conflicts are solved. Thanks @themurtazanazir mate

**Pull request merged**